### DISCUSSION TITLE
Remove UAT environment

### DISCUSSION THREAD

**Discussion by gidsg** [communitiesuk/funding-service-requests-for-comments#17] [General]
Title: Remove UAT environment
    Owner: Gideon Goldberg
    Collaborator(s):
    Created on: 2024-12-15
    Status: Draft 
    Finalised on: 

## Overview

Currently we have 4 deployed environments on pre-award 'Dev','Test','UAT' (User Acceptance Testing) and 'Production'. This is a request to remove 'UAT' as we believe it is no longer required and removing it would also speed up [Cycle Time](https://martinfowler.com/bliki/CycleTime.html) as it is currently a mandatory staging point prior to code going to Production.

## What is the current state?

Currently on pre-award we have 4 deployed environments:

| Environment name | Purpose |
|------------------|---------|
| Dev              | Testing branches |
| Test             | Reflects `main` branch, plus: Runs end-to-end tests |
| UAT              | Reflects `main` branch, plus: Runs end-to-end tests, Mirrors production scaling rules, Uses Live Notify key, Requires manual approval to promote to |
| Production       | Reflects `main` branch, plus: Used by end users, Requires manual approval to promote to |

### Illustration of current steps to deploy to production
```mermaid
flowchart LR
    A[Merge to main] --> B{Unit tests pass?}
    B --> C[Deploy to test]
    C --> D{E2E tests pass?}
    D --> E{Manual Approval}
    E --> F[Deploy to UAT]
    F --> G{E2E tests pass?}
    G --> H{Manual approval}
    H --> I[Deploy to production]
```

### Proposed steps to deploy production if RFC is approved
```mermaid
flowchart LR
    A[Merge to main] --> B{Unit tests pass?}
    B --> C[Deploy to test]
    C --> D{E2E tests pass?}
    D --> E{Manual Approval}
    E --> F[Deploy to production]    
```

## Why should we change?

As the same code is deployed to both Test and UAT there should be little if any difference between them, making this a mostly redundant step. 

This slows our [Cycle Time](https://martinfowler.com/bliki/CycleTime.html) by at least 15 minutes for the deployment plus end-to-end tests. This a minimum time as tests may fail due to flakiness or variance in configuration or database states and manual approval is required to promote to and from UAT, so may be further delayed should a developer not be attentive to promote their changes promptly.

As we use a Live Notify key on UAT this also increases our failed delivery rate, including sending test data to real domains such as `test.com` which is bad practice.

## How should we address this?

* Scale test to same size as prod
* Develop 'hot swap' mechanism to allow Notify key to be switched between Live and Test OR give all people who need to do demos access and training to use Notify API integration
* Use 'crash test dummy fund' for E2E tests to remove variance based on funds being open or closed. 
* Deploy 'fund-application-builder' and 'form-designer' to production
* Remove UAT

## What alternatives are we discarding?

* Retain UAT as-is
* Remove manual approval to promote to UAT (this should speed up deployments by not requiring manual developer intervention at that stage)

## Who will be affected?

* Will affect anyone that previously relied on the Live Notify Key for demos e.g. Product Managers doing show and tells on UAT
* Will affect anyone who previously used UAT for testing e.g. QA engineer
* Will affect any case where we relied on different environment config on UAT e.g. opening/closing rounds for test purposes

## Who will benefit?

* Developers will benefit from an improved experience as there will be fewer steps to deploying to production and fewer manual scripts to run due to having 1 fewer environment
* End users will benefit from faster fixes due to reduced Cycle Time
* Department will benefit from reduced infrastructure costs (~$1000 per month to run UAT)

## What are the key risks to manage or mitigate?

* Bugs occurs because Test environment does not have the same scaling rules as production (e.g. behavior with multiple instances)
    * Mitigation: Scale test to same size as prod
* Product Managers are unable to perform show and tells or independently test functionality due to lack of Live Notify Key
    * Mitigation: Develop 'hot swap' mechanism to allow Notify key to be switched between Live and Test OR give all people who need to do demos access to Notify API integration
* Bugs occur because configuration on 'test' does not match the configuration on 'production'
    * Mitigation: Use 'crash test dummy fund' for E2E tests to remove variance based on funds being open or closed, where fund specific config needs to be tested that can be done on Test
* Bugs occur because insufficient manual testing is done that previously would have been performed on UAT
    * Mitigation: If manual testing is required, deploy a branch to the 'dev' environment. Make use of E2E tests to increase confidence in code and 'feature flags' to deploy changes which should not yet be visible in production.
* Currently 'fund-application-builder' and 'form-designer' are using UAT as a temporary 'production-like' environment with different data, as the DNS for production has not been approved by TDA.
    * Mitigation: Do not remove UAT until 'fund-application-builder' and 'form-designer' are available on production

**Comment by wjrm500** [comment]
Probably no need for this comment given that the following is already under "What are the key risks to manage or mitigate?":

> Currently 'fund-application-builder' and 'form-designer' are using UAT as a temporary 'production-like' environment with different data, as the DNS for production has not been approved by TDA.
> Mitigation: Do not remove UAT until 'fund-application-builder' and 'form-designer' are available on production

But just to re-iterate the current functional importance of UAT from the perspective of Forms team

**Comment by samuelhwilliams** [comment]
I am in favour of removing the UAT environment, but want to note the following concerns/things we need to be aware of:

* We need to promote developer ownership of/responsibility for getting merged PRs into production so that changes don't pile up and deployments to prod become blocked by having to ask whether as-yet-unreleased changes are ready to go to production.
* We need to think about manual QA and when that happens. If it can happen in dev pre-merge (and `dev` is a fair representation of reality), then that feels best. If it needs to happen in `test`, then we should make sure that QA is ready to do that testing as soon as it's been merged and deployed to that environment - again to prevent blocking other people.
* We may need to have support for some ephemeral environments in case the `dev` env starts to become a contested area for testing pre-merge.

Stuff to consider for the future that could mitigate the above is removing the manual approval step before going to production. I think we need to improve our automated smoke/e2e/performance test suite against the `test` environment before we could do that, but I think that would be a good aim.

**Reply by wjrm500** [reply]
Agreed with all of these considerations, I was wondering about your third point earlier this week - specifically should we introduce pull previews in FAB to avoid sharing of dev environment. I was quite keen although conscious of resource usage, Nuwan mentioned concerns over replicating AWS-hosted environment effectively in pull previews

**Reply by gidsg** [reply]
> specifically should we introduce pull previews in FAB

@wjrm500 this is certainly something you could spike, I think some challenges are around auth:
* We cannot use Entra ID/Authenticator as that relies on fixed callback URLs. We would probably want this at least behind basic auth at a minimum
* Data population (as each preview environment starts fresh, we'd need some way of 'seeding' the environment if we wanted it to have populated data.)

>  conscious of resource usage

As the thing we have prototyped uses Lightsail instances the cost is mimimal

**Reply by samuelhwilliams** [reply]
> Data population (as each preview environment starts fresh, we'd need some way of 'seeding' the environment if we wanted it to have populated data.)

This shouldn't be a huge issue because we seed the local DB correctly from scratch now.

The auth thing is maybe a question - but also we have that hacky debug user support that could potentially be used to bypass the login flow (although there's obviously a bit of sadness around that).

Would be up for someone spiking this - might be easiest to do once pre-award-frontend and pre-award-stores have been combined. Would also request that anyone doing this do a kickoff to gather 'requirements'/scope from across the dev team

**Comment by srh-sloan** [comment]
Agree that UAT is currently a redundant step in our pipeline, from a dev perspective, as we don't rely on it for anything other than testing with multiple instances. That will be mitigated by scaling test to the same level of production, which is already listed.

Also agree with Sam's comments above about making sure developers are prompt in deploying merged changes to production, and only merging when they are ready, to avoid delays and 'big bang' releases. We need to think about this anyway as we move towards a monolith, so this will be another reason to improve our workflows around this.

**Comment by nuwan-samarasinghe** [comment]
Can we standardise our all environments to have same configurations I see some time slight changes needed for the prod env..
also may be better if we can change the environments into this way may be just thought of my self 

test -> rename into dev
uat -> rename into test (I am suggesting that this should mimic the prod env)
prod same

**Comment by wjrm500** [comment]
Just rekindling this one now that FAB is in prod - happy to wave goodbye to UAT 👋
